src/VisualStudio/Core/Def/Implementation/CodeLens/CodeLensCallbackListener.cs

```diff
@@ -61,35 +60,16 @@ internal class CodeLensCallbackListener : ICodeLensCallbackListener, ICodeLensCo
             _workspace = workspace;
         }
 
-        public async Task<string?> TryGetHostGroupIdAsync(CancellationToken cancellationToken)
+        public async Task<ImmutableDictionary<Guid, string>> GetProjectVersionsAsync(CancellationToken cancellationToken)
         {
-            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
-            if (client == null)
+            var builder = ImmutableDictionary.CreateBuilder<Guid, string>();
+            foreach (var project in _workspace.CurrentSolution.Projects)
             {
-                return null;
-            }
-
-            return client.ClientId;
-        }
-
-        public Task<string?> TryGetServiceNameAsync(CancellationToken cancellationToken)
-        {
-            if (!RemoteHostOptions.IsUsingServiceHubOutOfProcess(_workspace.Services))
-                return SpecializedTasks.Null<string>();
-
-            var isRemoteHost64Bit = RemoteHostOptions.IsServiceHubProcess64Bit(_workspace.Services);
-            return Task.FromResult<string?>(new RemoteServiceName(WellKnownServiceHubService.CodeAnalysis).ToString(isRemoteHost64Bit));
-        }
-
-        public List<Guid>? GetDocumentId(Guid projectGuid, string filePath, CancellationToken cancellationToken)
-        {
-            if (TryGetDocument(_workspace.CurrentSolution, projectGuid, filePath, out var document))
-            {
-                var documentId = document.Id;
-                return new List<Guid>(2) { documentId.ProjectId.Id, documentId.Id };
+                var projectVersion = await project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
❔ How expensive is `GetDependentVersionsAsync`?

1. Do we need to implement a `WorkspaceChanged` throttle here, or just make the calls to get the version?
2. Do we need to add an argument to only get the versions for projects we currently have data points for?